File: block_af_alg.c
/* 
 * Standalone eBPF program to block AF_ALG sockets
 * Compile with: clang -O2 -g -target bpf -D__TARGET_ARCH_x86 -c block_af_alg.c -o block_af_alg.o
 * Load with: sudo bpftool prog loadall block_af_alg.o /sys/fs/bpf/block_af_alg autoattach
 * Note: This is an LSM program (not XDP/kprobe). Kernel must support BPF LSM.
 */

#include <linux/bpf.h>
#include <linux/socket.h>
#include <bpf/bpf_helpers.h>
#include <bpf/bpf_tracing.h>

#define AF_ALG 38
#define EPERM 1

/* Map to store statistics */
struct {
    __uint(type, BPF_MAP_TYPE_ARRAY);
    __uint(max_entries, 10);
    __type(key, __u32);
    __type(value, __u64);
} af_alg_stats SEC(".maps");

/* LSM hook: socket_create
 * Blocks socket() calls when family == AF_ALG.
 * Returning negative errno from LSM hooks is the supported deny mechanism.
 */
SEC("lsm/socket_create")
int BPF_PROG(trace_socket_create, int family, int type, int protocol, int kern) {
    
    if (family == AF_ALG) {
        __u32 idx = 0;
        __u64 *val = bpf_map_lookup_elem(&af_alg_stats, &idx);
        if (val) {
            __sync_fetch_and_add(val, 1);
        }
        
        // Log the event - it will go to /sys/kernel/tracing/trace_pipe
        bpf_printk("BLOCKED: AF_ALG socket attempt, pid=%d uid=%d",
                   bpf_get_current_pid_tgid() >> 32,
                   bpf_get_current_uid_gid() & 0xFFFFFFFF);

        // Deny the socket creation.
        return -EPERM;
    }

    return 0;
}

char LICENSE[] SEC("license") = "GPL";

File: block-af-alg.py
#!/usr/bin/env python3
"""
BCC eBPF program to BLOCK AF_ALG socket creation via LSM hook.
AF_ALG is address family 38, used for kernel crypto socket interface.

Uses LSM_PROBE(socket_create) which allows returning -EPERM to deny.
Requires:
  - CONFIG_BPF_LSM=y
  - 'bpf' in /sys/kernel/security/lsm
  - BCC 0.26+
  - Kernel 5.7+
"""

from bcc import BPF
import sys
import signal

bpf_source = """
#include <linux/socket.h>

#define AF_ALG 38

BPF_ARRAY(block_count, u64, 1);

LSM_PROBE(socket_create, int family, int type, int protocol, int kern) {
    if (family == AF_ALG) {
        u32 idx = 0;
        u64 *count = block_count.lookup(&idx);
        if (count) {
            __sync_fetch_and_add(count, 1);
        }
        bpf_trace_printk("BLOCKED: AF_ALG socket attempt, pid=%u\\n",
                         bpf_get_current_pid_tgid() >> 32);
        return -1;
    }
    return 0;
}
"""

def main():
    print("[*] Loading AF_ALG socket blocker eBPF program (LSM hook - actual blocking)")
    
    # Create BPF object
    try:
        b = BPF(text=bpf_source)
    except Exception as e:
        print(f"[!] Error loading BPF: {e}")
        print("[!] Ensure CAP_SYS_ADMIN, CAP_SYS_RESOURCE, CAP_PERFMON are available")
        print("[!] Run as: sudo python3 block-af-alg.py")
        sys.exit(1)
    
    # LSM_PROBE is auto-attached by BCC.
    print("[+] Attached LSM hook: socket_create")
    
    # Print trace output
    print("[+] Blocking AF_ALG socket attempts (Ctrl-C to stop)...")
    print("[+] Output from trace_pipe:")
    
    def signal_handler(sig, frame):
        # Print statistics
        print("\n[*] Unloading...")
        try:
            count_val = b["block_count"][0].value
            print(f"[+] Total AF_ALG socket() calls blocked (denied): {count_val}")
        except Exception:
            pass
        sys.exit(0)
    
    signal.signal(signal.SIGINT, signal_handler)
    
    # Read trace output
    b.trace_print()

if __name__ == "__main__":
    main()
